core/drivers/amd/mailbox_driver.c

```diff
@@ -0,0 +1,216 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2024-2025, Advanced Micro Devices, Inc. All rights reserved.
+ *
+ */
+#include <drivers/amd/mailbox_driver.h>
+#include <initcall.h>
+#include <kernel/delay.h>
+#include <kernel/panic.h>
+#include <mailbox_private.h>
+#include <malloc.h>
+#include <mm/core_mmu.h>
+#include <platform_config.h>
+#include <string.h>
+#include <tee/cache.h>
+
+static struct ipi_info_t ipi_info;
+
+static TEE_Result mailbox_call(enum mailbox_api ipi_api, uint32_t blocking)
+{
+	struct thread_smc_args args = {
+		.a0 = PLAT_SMC_SIP_IPI | ipi_api,
+		.a1 = reg_pair_to_64(0, ipi_info.local),
+		.a2 = reg_pair_to_64(0, ipi_info.remote),
+		.a3 = reg_pair_to_64(0, blocking),
+	};
+
+	thread_smccc(&args);
+
+	if (args.a0) {
+		EMSG("Mailbox Call returned with error 0x%lx", args.a0);
+		return TEE_ERROR_GENERIC;
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result mailbox_open(uint32_t remote_id, size_t payload_size)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	paddr_t req_addr = 0, rsp_addr = 0;
+
+	ipi_info.remote = remote_id;
+
+	req_addr = (ipi_info.buf + (remote_id * IPI_OFFSET_MULTIPLIER) +
+		    IPI_REQ_OFFSET);
+
+	rsp_addr = (ipi_info.buf + (remote_id * IPI_OFFSET_MULTIPLIER) +
+		    IPI_RESP_OFFSET);
+
+	ipi_info.req = core_mmu_add_mapping(MEM_AREA_RAM_SEC,
+					    req_addr,
+					    payload_size);
+
+	ipi_info.rsp = core_mmu_add_mapping(MEM_AREA_RAM_SEC,
+					    rsp_addr,
+					    payload_size);
+	if (!ipi_info.req || !ipi_info.rsp)
+		panic();
+
+	mutex_lock(&ipi_info.lock);
+
+	res = mailbox_call(IPI_MAILBOX_OPEN, IPI_BLOCK);
+	if (res != TEE_SUCCESS)
+		EMSG("Failed to open Mailbox for remote id %"PRIu32, remote_id);
+
+	mutex_unlock(&ipi_info.lock);
+
+	return res;
+}
+
+TEE_Result mailbox_release(uint32_t remote_id, size_t payload_size)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	if (ipi_info.remote != remote_id) {
+		EMSG("Mailbox not available for remote id %"PRIu32, remote_id);
+		return res;
+	}
+
+	mutex_lock(&ipi_info.lock);
+
+	res = mailbox_call(IPI_MAILBOX_RELEASE, IPI_BLOCK);
+	if (res != TEE_SUCCESS) {
+		EMSG("Mailbox release failed for remote id %"PRIu32, remote_id);
+		goto out;
+	}
+
+	res = core_mmu_remove_mapping(MEM_AREA_RAM_SEC,
+				      ipi_info.req,
+				      payload_size);
+	if (res != TEE_SUCCESS) {
+		EMSG("Failed to remove response mapping for remote id %"PRIu32,
+		     remote_id);
+		goto out;
+	}
```

Comment:
Here and below, `ipi_info.req` and `ipi_info.rsp` should be reset to NULL to avoid pointing to an unmapped address or more dramatically another physical location.